src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -111685,5 +111685,184 @@ void M(object? o1)
+    {{
+        locationNonNull = new object();
+        Interlocked.CompareExchange(ref {location}, {value}, {comparand});
+        _ = {location}.ToString();
+    }}
+}}
+    ";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(isLocationNonNull
+                ? Array.Empty<DiagnosticDescription>()
+                : new[]
+                {
+                    // (18,13): warning CS8602: Dereference of a possibly null reference.
+                    //         _ = {location}.ToString();
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, location).WithLocation(18, 13)
+                });
+
+            source = $@"
+#pragma warning disable 0436
+using System.Threading;
+namespace System.Threading
+{{
+	public static class Interlocked
+    {{
+        public static T CompareExchange<T>(ref T location, T value, T comparand) where T : class? {{ return location; }}
+    }}
+}}
+
+class C
+{{
+    void M(string? locationNonNull, string? locationMaybeNull, string comparandNonNull, string? comparandMaybeNull, string valueNonNull, string? valueMaybeNull)
+    {{
+        locationNonNull = ""hello"";
+        Interlocked.CompareExchange(ref {location}, {value}, {comparand});
+        _ = {location}.ToString();
+    }}
+}}
+    ";
+            comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(isLocationNonNull
+                ? Array.Empty<DiagnosticDescription>()
+                : new[]
+                {
+                    // (18,13): warning CS8602: Dereference of a possibly null reference.
+                    //         _ = {location}.ToString();
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, location).WithLocation(18, 13)
+                });
+
+            source = $@"
+class C
+{{
+    void M(object? locationNonNull, object? locationMaybeNull, object comparandNonNull, object? comparandMaybeNull, object valueNonNull, object? valueMaybeNull)
+    {{
+        locationNonNull = new object();
+        if ({location} == {comparand})
+        {{
+            {location} = {value};
+        }}
+        _ = {location}.ToString();
+    }}
+}}
+    ";
+            comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(isLocationNonNull
+                ? Array.Empty<DiagnosticDescription>()
+                : new[]
+                {
+                    // (11,13): warning CS8602: Dereference of a possibly null reference.
+                    //         _ = {location}.ToString();
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, location).WithLocation(11, 13)
+                });
+        }
+
+        [Fact]
+        [WorkItem(36911, "https://github.com/dotnet/roslyn/issues/36911")]
+        public void CompareExchange_BadOverload()
```

Comment:
>Bad [](start = 36, length = 3)

Perhaps "OtherOverload", here and test below, because the overloads are not recognized but valid.